apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts

```diff
@@ -0,0 +1,235 @@
+    if (options?.convertLineEndings) {
+      contents = Text.convertTo(contents, options.convertLineEndings);
+    }
+
+    return contents;
+  };
+
+  public readFileToBuffer: (filePath: string) => Buffer = (filePath: string) => {
+    return this._withCaching(filePath, FileSystem.readFileToBuffer, this._readFileCache);
+  };
+
+  public copyFileAsync: (options: IFileSystemCopyFileOptions) => Promise<void> = async (
+    options: IFileSystemCopyFileOptions
+  ) => {
+    await FileSystem.copyFileAsync(options);
+    this._invalidateCacheEntry(options.destinationPath);
+  };
+
+  public deleteFile: (filePath: string, options?: IFileSystemDeleteFileOptions | undefined) => void = (
+    filePath: string,
+    options?: IFileSystemDeleteFileOptions | undefined
+  ) => {
+    const cachedError: Error | undefined = this._statsCache.get(filePath)?.error;
+    if (!cachedError || !FileSystem.isFileDoesNotExistError(cachedError)) {
+      FileSystem.deleteFile(filePath);
+      this._invalidateCacheEntry(filePath);
+    } else if (options?.throwIfNotExists) {
+      throw cachedError;
+    }
+  };
+
+  public createHardLinkAsync: (options: IFileSystemCreateLinkOptions) => Promise<void> = async (
+    options: IFileSystemCreateLinkOptions
+  ) => {
+    await FileSystem.createHardLinkAsync(options);
+    this._invalidateCacheEntry(options.newLinkPath);
+  };
+
+  public getRealPath: (linkPath: string) => string = (linkPath: string) => {
+    return this._withCaching(linkPath, FileSystem.getRealPath, this._realPathCache);
+  };
+
+  public readFolderFilesAndDirectories: (folderPath: string) => IReadFolderFilesAndDirectoriesResult = (
+    folderPath: string
+  ) => {
+    return this._withCaching(
+      folderPath,
+      (path: string) => {
+        // TODO: Replace this with a FileSystem API
+        const folderEntries: fs.Dirent[] = fs.readdirSync(path, { withFileTypes: true });
+        return this._sortFolderEntries(folderEntries);
+      },
+      this._readFolderCache
+    );
+  };
+
+  public createHardLinkExtendedAsync: (options: ICreateHardLinkExtendedOptions) => Promise<boolean> = async (
+    options: ICreateHardLinkExtendedOptions
+  ) => {
+    try {
+      await this.createHardLinkAsync(options);
+      return true;
+    } catch (error) {
+      if (error.code === 'EEXIST') {
+        if (options.preserveExisting) {
+          return false;
+        }
+
+        this.deleteFile(options.newLinkPath);
+      } else if (FileSystem.isNotExistError(error)) {
+        await this.ensureFolderAsync(nodeJsPath.dirname(options.newLinkPath));
+      } else {
+        throw error;
+      }
+
+      await this.createHardLinkAsync(options);
+      return true;
+    }
+  };
+
+  private _sortFolderEntries(folderEntries: fs.Dirent[]): IReadFolderFilesAndDirectoriesResult {
+    // TypeScript expects entries sorted ordinally by name
+    // In practice this might not matter
+    folderEntries.sort((a, b) => {
```

Comment:
Compare with [Sort.compareByValue()](https://github.com/microsoft/rushstack/blob/0f193ac7afcd2527370ebaa342299bb31be54e27/libraries/node-core-library/src/Sort.ts#L34) for some possible edge cases